### PR TITLE
Port unfactored Roslyn analyzer rules and add unit tests

### PR DESCRIPTION
**Pull request by mavasani**
1. DoNotMixAttributesFromDifferentVersionsOfMEFAnalyzer
2. PartsExportedWithMEFv2MustBeMarkedAsShared
3. AvoidUsingCrefTagsWithAPrefixAnalyzer
4. DoNotCallToImmutableArrayOnAnImmutableArrayValueAnalyzer
5. DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyAnalyzer
6. DoNotCreateTasksWithoutPassingATaskSchedulerAnalyzer

Fixes https://github.com/dotnet/roslyn-analyzers/issues/954

### CODE REVIEW

**Comment by mavasani** [comment]
TODO: Add unit tests for DoNotMixAttributesFromDifferentVersionsOfMEFAnalyzer. Will do it as a separate commit/PR

**Comment by mavasani** [comment]
Tagging @dotnet/roslyn-analysis @srivatsn @basoundr @lgolding for review.

File: src/Desktop.Analyzers/UnitTests/Desktop.Analyzers.UnitTests.csproj [hunk 1]
```diff
@@ -128,6 +128,9 @@
     <Compile Include="TypesShouldNotExtendCertainBaseTypesTests.cs" />
     <Compile Include="TypesShouldNotExtendCertainBaseTypesTests.Fixer.cs" />
   </ItemGroup>
+  <ItemGroup>
+    <Service Include="{82A7F48D-3B50-4B1E-B82E-3ADA8210C358}" />
```
Comment (srivatsn) [inline src/Desktop.Analyzers/UnitTests/Desktop.Analyzers.UnitTests.csproj]: Any idea what these are?

File: src/Microsoft.Composition.Analyzers/Core/DoNotMixAttributesFromDifferentVersionsOfMEF.cs [hunk 2]
```diff
@@ -1,15 +1,19 @@
...
 namespace Microsoft.Composition.Analyzers
 {
     /// <summary>
     /// RS0006: Do not mix attributes from different versions of MEF
     /// </summary>
-    public abstract class DoNotMixAttributesFromDifferentVersionsOfMEFAnalyzer : DiagnosticAnalyzer
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class DoNotMixAttributesFromDifferentVersionsOfMEFAnalyzer : DiagnosticAnalyzer
     {
         internal const string RuleId = "RS0006";
```
Comment (srivatsn) [inline src/Microsoft.Composition.Analyzers/Core/DoNotMixAttributesFromDifferentVersionsOfMEF.cs]: Can you give them some CA ids? Mike has a list of all used FxCop ids that we can use to get the next free one.

File: src/Microsoft.Composition.Analyzers/Core/PartsExportedWithMEFv2MustBeMarkedAsShared.cs [hunk 3]
```diff
@@ -1,15 +1,18 @@
...
 namespace Microsoft.Composition.Analyzers
 {
     /// <summary>
     /// RS0023: Parts exported with MEFv2 must be marked as Shared
     /// </summary>
-    public abstract class PartsExportedWithMEFv2MustBeMarkedAsSharedAnalyzer : DiagnosticAnalyzer
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class PartsExportedWithMEFv2MustBeMarkedAsSharedAnalyzer : DiagnosticAnalyzer
     {
         internal const string RuleId = "RS0023";
```
Comment (srivatsn) [inline src/Microsoft.Composition.Analyzers/Core/PartsExportedWithMEFv2MustBeMarkedAsShared.cs]: Same comment about CA ids.

**Comment by srivatsn** [comment]
>                                                                              description: s_localizableDescription,

true

---
Refers to: src/System.Runtime.Analyzers/Core/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.cs:37 in e341782. [](commit_id = e341782a8796570a70254c6083af36a5900acf10, deletion_comment = False)

File: src/System.Threading.Tasks.Analyzers/Core/DoNotCreateTasksWithoutPassingATaskScheduler.cs [hunk 4]
```diff
@@ -25,13 +28,65 @@ public abstract class DoNotCreateTasksWithoutPassingATaskSchedulerAnalyzer : Dia
                                                                              DiagnosticSeverity.Warning,
                                                                              isEnabledByDefault: false,
```
Comment (srivatsn) [inline src/System.Threading.Tasks.Analyzers/Core/DoNotCreateTasksWithoutPassingATaskScheduler.cs]: true

**Comment by srivatsn** [comment]
For these rules, please reopen their Github issues and mark them as "Needs-Fixer" as well.

**Comment by srivatsn** [comment]
Thanks for moving all of these over!

**Comment by srivatsn** [comment]
:+1:

File: src/Desktop.Analyzers/UnitTests/Desktop.Analyzers.UnitTests.csproj (hunk 1 above)
Comment (mavasani) [inline src/Desktop.Analyzers/UnitTests/Desktop.Analyzers.UnitTests.csproj]: No idea how these got added - will revert

File: src/Microsoft.Composition.Analyzers/Core/DoNotMixAttributesFromDifferentVersionsOfMEF.cs (hunk 2 above)
Comment (mavasani) [inline src/Microsoft.Composition.Analyzers/Core/DoNotMixAttributesFromDifferentVersionsOfMEF.cs]: I'll do this in a separate change.

**Comment by mavasani** [comment]
I have pushed another commit addressing PR feedback and adding unit tests for DoNotMixAttributesFromDifferentVersionsOfMEFAnalyzer

**Comment by advancedrei** [comment]
FYI, the DoNotCreateTasks.... analyzer does not appear to function in VS2015 Update 3. I'm getting the error: 

`Analyzer 'System.Threading.Tasks.Analyzers.DoNotCreateTasksWithoutPassingATaskSchedulerAnalyzer' threw an exception of type 'System.InvalidOperationException' with message 'Feature 'IOperation' is disabled.'.`

**Comment by srivatsn** [comment]
The analyzer uses an experimental feature called IOperation. To enable that you'll need to add <Features>IOperation</Features> to your project file.

**Comment by mavasani** [comment]
Specifically, add a new Property `<Features>IOperation</Features>` to any PropertyGroup in your project file.

**Pull request merged**